Initial Linux platform port of lwpa

samkearney opened the pull request.

samkearney committed:
Merge branch 'develop' into linux-port

samkearney committed:
Merge branch 'develop' into linux-port

samkearney committed:
Checking in current WIP not compiling

samkearney committed:
More WIP

samkearney committed:
Still working on netints

samkearney committed:
Push up WIP to continue on Windows

samkearney committed:
Get things working on Windows again

samkearney committed:
Finish up some API changes on Windows, moving back to Linux

samkearney committed:
Routing table now working.

samkearney committed:
Improve netint and example. Gateway code does not quite work as expected yet - need to troubleshoot

samkearney committed:
Add unit test for is_loopback function

samkearney committed:
Fix gcc warnings

samkearney committed:
Implement some of the basic socket functions

samkearney committed:
Finish up Linux netints

samkearney committed:
Cleanup windows side

samkearney committed:
Working through some socket code, WIP

samkearney committed:
Finish up socket options

samkearney committed:
Lots of progress, only 4 failing unit tests now.

samkearney committed:
Windows netint fix

samkearney committed:
Merge branch 'linux-port' of https://github.com/ETCLabs/lwpa into linux-port

samkearney committed:
Windows changes in progress

samkearney committed:
Finish fixing windows for changed APIs

samkearney committed:
Threads implemented. Linux - all unit tests passing

samkearney committed:
Cleanup, fix some warnings, add pipeline step

samkearney committed:
Small optimization to linux threads

samkearney committed:
And fix Windows a final time

samkearney committed:
Add UUID dependency to Linux pipeline script

samkearney committed:
Troubleshooting Linux pipeline

nwagner commented on external/fff/CMakeLists.txt:
```diff
@@ -1,6 +1,8 @@
 if(NOT TARGET meekrosoft::fff)
   add_library(meekrosoft::fff INTERFACE IMPORTED GLOBAL)
```
what the heck is meekrosoft, other than a vague homonym?

samkearney replied:
A Github username: https://github.com/meekrosoft/fff

nwagner commented on include/lwpa/inet.h:
```diff
@@ -72,47 +76,47 @@ typedef struct LwpaIpAddr
-    (lwpa_ip_ptr)->type = kLwpaIpTypeV4;        \
-    (lwpa_ip_ptr)->addr.v4 = val;               \
+#define LWPA_IP_SET_V4_ADDRESS(lwpa_ip_ptr, val) \
+  do                                             \
```
What's the point of the do/while(0) again?  Is it for scoping?

samkearney replied:
It's a C macro trick to make them able to be used wherever functions are. https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for

nwagner commented on src/lwpa/inet.c:
```diff
@@ -0,0 +1,361 @@
+    }
+    else if (ip1->type == kLwpaIpTypeV4)
+    {
+      return ((int)LWPA_IP_V4_ADDRESS(ip1) - (int)LWPA_IP_V4_ADDRESS(ip2));
```
Wouldn't this comparison be wrong when the first octet of at least one address is >= 128?

samkearney replied:
Whoa, you ain't kidding! Who wrote this crap?

samkearney replied:
The correct algorithm should be
```
return (LWPA_IP_V4_ADDRESS(ip1) > LWPA_IP_V4_ADDRESS(ip2)) - (LWPA_IP_V4_ADDRESS(ip1) < LWPA_IP_V4_ADDRESS(ip2));
```
Will implement this.

nwagner replied:
I think there may be a couple of other comparisons you do (maybe with OS indexes or platform socket handles) that could have a similar issue if the indexes somehow have the high bit set.

nwagner commented on src/os/linux/lwpa/os_thread.c:
```diff
@@ -0,0 +1,70 @@
+  return res;
+}
+
+bool lwpa_thread_stop(lwpa_thread_t* id)
```
I'm a little confused -- this doesn't actually stop the thread, it just waits for it to die.

samkearney replied:
That's true - would you expect this function to wrap a `pthread_cancel()` call instead?

nwagner replied:
No, I'd rather the join be wrapped.  On windows, I see you're doing a WaitForSingleObject, which is similar.  Maybe change the name of the function -- e.g lwpa_wait_for_thread_stop?

samkearney replied:
Maybe it should be `lwpa_thread_join()`? That's the nomenclature used by POSIX and both the C++ and C standard libraries for threads.

nwagner replied:
That would be ideal.  I only hesitated because Windows doesn't use that name -- but that's a minority that documentation will cover.

nwagner commented on src/os/windows/lwpa/os_netint.c:
```diff
@@ -20,289 +20,335 @@
 }
 
-size_t lwpa_netint_get_interfaces(LwpaNetintInfo* netint_arr, size_t netint_arr_size)
+lwpa_error_t lwpa_netint_get_interface_for_dest(const LwpaIpAddr* dest, LwpaNetintInfo* netint)
```
Shouldn't we call GetBestInterfaceEx instead?  Or maybe not, given our last discussion..

samkearney replied:
No, I think you are right and we should be.

nwagner commented on tests/unit/test_inet.cpp:
```diff
@@ -19,51 +19,266 @@
   LwpaIpAddr v4_less;
-  lwpaip_set_v4_address(&v4_less, 0x01020303);
+  LWPA_IP_SET_V4_ADDRESS(&v4_less, 0x01020303);
 
```
You're not testing the case I mentioned earlier (the first octet has the high bit set)

samkearney replied:
Will add

nwagner commented on src/os/linux/lwpa/os_netint.c:
```diff
@@ -17,4 +17,682 @@
+
+/*************************** Function definitions ****************************/
+
+lwpa_error_t lwpa_netint_init()
```
I feel like there's some overlap between this and windows.  Could there be some shared code to get rid of boilerplate?  Or is there enough differences that it isn't worth it..

samkearney replied:
I'll take a second look. Last time I looked at it it was a bit too different.

nwagner reviewed (commented):
Looking really good!  Have you run these through static analyzers yet?

samkearney committed:
Fix a static analysis error, fix lwpa_ip_cmp() and add test

samkearney committed:
Improve thread API

samkearney committed:
Some updates from code review (not done yet):
  - Create a common module for netint and share API function code
  - Use GetBestInterfaceEx on Windows for
    lwpa_netint_get_interface_for_destination()
  - Move sockaddr conversion and lwpa_inet_xtox() functions to lwpa_inet
  - Touch up some documentation
  - Update changelog

samkearney committed:
And fix the Linux side.
